### PR TITLE
First deliverable: RxJava, Singleton, MVVM Pattern and RecyclerView Pagination

### PR DESCRIPTION
**Pull request by ponchOlvera**
Implemented MVVM pattern.
Implemented a Singleton pattern in the Retrofit instance to prevent multiple instances.
Use RxJava to return an Observable when executing a HTTP request.
Implemented a simple pagination in the RecyclerView and populate Marvel characters in the list.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -30,20 +30,38 @@ android {
...
+  implementation 'com.google.android.material:material:1.4.0'
+  implementation 'androidx.constraintlayout:constraintlayout:2.1.2'
+
+  // RecyclerView
+  implementation "androidx.recyclerview:recyclerview:$rootProject.recyclerViewVersion"
+  implementation "androidx.paging:paging-runtime-ktx:$paging_version"
+
+  //Glide
+  implementation 'com.github.bumptech.glide:glide:4.12.0'
+  annotationProcessor 'com.github.bumptech.glide:compiler:4.12.0'
```
Comment (pestrada) [inline app/build.gradle]: The suggestion here is to use variables for setting any library dependencies. Similar to what you did with `paging_version`

File: app/src/main/java/com/wizeline/heroes/HeroesServices.kt [hunk 2]
```diff
@@ -11,5 +15,22 @@ interface HeroesServices {
         @Query("ts") ts: String,
         @Query("apikey") apikey: String,
         @Query("hash") hash: String,
-    ): Call<Characters>
+        @Query("offset") offset: Int,
+    ): Single<Characters>
+
+    @GET("characters/{characterId}/comics")
+    fun comicsByCharacter(
+        @Query("ts") ts: String,
```
Comment (pestrada) [inline app/src/main/java/com/wizeline/heroes/HeroesServices.kt]: Suggestion -> Let's use more descriptive names for variables. Example instead of `ts` use `timeStamp` or whatever the real name is.
```
@Query("ts") timeStamp: String,
```

File: app/src/main/java/com/wizeline/heroes/NetworkClient.kt [hunk 3]
```diff
@@ -1,14 +1,26 @@
...
+
+    private var retrofitInstance: Retrofit? = null
+
+    init {
+        if (retrofitInstance == null) createRetrofitInstance()
+    }
 
     fun getServices(): HeroesServices {
-        return Retrofit.Builder().apply {
+        return retrofitInstance!!.create(HeroesServices::class.java)
```
Comment (pestrada) [inline app/src/main/java/com/wizeline/heroes/NetworkClient.kt]: We should avoid the operator `!!` because it can throw a runtime exception.
```suggestion
       if (retrofitInstance == null) createRetrofitInstance()
       return retrofitInstance.create(HeroesServices::class.java)
```

File: app/src/main/java/com/wizeline/heroes/NetworkClient.kt [hunk 4]
```diff
@@ -1,14 +1,26 @@
...
+    }
+
+    private fun createRetrofitInstance(){
+        retrofitInstance = Retrofit.Builder().apply {
             baseUrl("https://gateway.marvel.com/v1/public/")
```
Comment (pestrada) [inline app/src/main/java/com/wizeline/heroes/NetworkClient.kt]: In general, we can use a single source of truth for defining API endpoints. Something like:
```
class Endpoint {
  const val BASE_URL = "https://gateway.marvel.com/v1/public/"
  consts val COMICS_URL = "characters/{characterId}/comics"
  // etc.
}
```

File: app/src/main/java/com/wizeline/heroes/models/Character.kt [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import com.wizeline.heroes.Thumbnail
+
+data class Character(
+    @SerializedName("id") val id: Int,
+    @SerializedName("name") val name: String,
+    @SerializedName("description") val description: String,
+    @SerializedName("modified") val modified: String,
+    @SerializedName("comics") val comics: ComicList,
+    @SerializedName("thumbnail") val thumbnail: Thumbnail,
+    @SerializedName("resourceURI") val resourceURI: String
```
Comment (pestrada) [inline app/src/main/java/com/wizeline/heroes/models/Character.kt]: When the serialized name is exactly the same as the variable name. You can omit the serialized name.
`val name: String,`

File: app/src/main/java/com/wizeline/heroes/models/ComicData.kt [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package com.wizeline.heroes.models
+
+import com.google.gson.annotations.SerializedName
+
+data class ComicData(
+    @SerializedName("code") val code: Int
+) {
```
Comment (pestrada) [inline app/src/main/java/com/wizeline/heroes/models/ComicData.kt]: Let's omit the curly braces if not needed.

File: app/src/main/java/com/wizeline/heroes/ui/ExtensionFunctions.kt [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+import android.app.AlertDialog
+import android.content.Context
+import android.net.ConnectivityManager
+import android.net.NetworkCapabilities
+import android.os.Build
+import android.view.LayoutInflater
+import androidx.fragment.app.Fragment
+import com.wizeline.heroes.R
+
+object ExtensionFunctions {
```
Comment (pestrada) [inline app/src/main/java/com/wizeline/heroes/ui/ExtensionFunctions.kt]: Maybe rename to `FragmentExtensions.kt` in order to convey their intent.

File: app/src/main/java/com/wizeline/heroes/ui/ExtensionFunctions.kt [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        progressDialog?.dismiss()
+    }
+
+    fun Fragment.isInternetAvailable(context: Context): Boolean {
+        var result = false
+        val connectivityManager =
+            context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            val networkCapabilities = connectivityManager.activeNetwork ?: return false
+            val actNw =
```
Comment (pestrada) [inline app/src/main/java/com/wizeline/heroes/ui/ExtensionFunctions.kt]: let's rename to `activeNetworkCapabilities` for a clearer name.

File: app/src/main/java/com/wizeline/heroes/ui/abstract/PaginationScrollListener.kt [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+        super.onScrolled(recyclerView, dx, dy)
+
+        val visibleItemCount = layoutManager.childCount
+        val totalItemCount = layoutManager.itemCount
+        val firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition()
+
+        if (!isLoading() && !isLastPage()) {
+            if (visibleItemCount + firstVisibleItemPosition >= totalItemCount && firstVisibleItemPosition >= 0) {
+                loadMoreItems()
+            }//                    && totalItemCount >= ClothesFragment.itemsCount
```
Comment (pestrada) [inline app/src/main/java/com/wizeline/heroes/ui/abstract/PaginationScrollListener.kt]: let's remove commented code.

File: app/src/main/java/com/wizeline/heroes/ui/adapters/CharacterRecyclerViewAdapter.kt [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.*
+import kotlin.collections.ArrayList
+
+class CharacterRecyclerViewAdapter(
+    private val list: ArrayList<Character>,
+    private val context: Context
+) :
+    RecyclerView.Adapter<CharacterRecyclerViewAdapter.ViewHolder>() {
+    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
+        val imgCharacter: ImageView = view.findViewById(R.id.imgCharacter)
```
Comment (pestrada) [inline app/src/main/java/com/wizeline/heroes/ui/adapters/CharacterRecyclerViewAdapter.kt:24]: Can we investigate a bit more why ViewBinding was not working here? For the next deliverable.

File: app/src/main/java/com/wizeline/heroes/ui/adapters/CharacterRecyclerViewAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+            context.getString(R.string.available_comics),
+            character.comics.available
+        )
+
+        //Set image
+        Glide.with(context)
+            .load(imagePath)
+            .centerCrop()
+            .placeholder(R.drawable.ic_superhero_placeholder)
+            .into(holder.imgCharacter)
```
Comment (pestrada) [inline app/src/main/java/com/wizeline/heroes/ui/adapters/CharacterRecyclerViewAdapter.kt]: suggestion, move this code to inside ViewHolder on a function:
```
ViewHolder {
   fun bind(character: Character) {
     holder.tvCharacterName.text = character.name
        holder.tvCharacterDescription.text = description
        // etc
   }
}
```

File: app/src/main/java/com/wizeline/heroes/ui/adapters/CharacterRecyclerViewAdapter.kt [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        val view = LayoutInflater.from(context)
+            .inflate(R.layout.recycler_view_character_item, parent, false)
+        return ViewHolder(view)
+    }
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        val character = list.get(position)
+        val description =
+            if (character.description.isEmpty()) context.getString(R.string.description_not_available) else character.description
```
Comment (pestrada) [inline app/src/main/java/com/wizeline/heroes/ui/adapters/CharacterRecyclerViewAdapter.kt]: It's a good idea to move this kind of presentation logic to a separate class, in order to let the ViewHolder just render things. Some alternatives can be:
1.- Using a Presentation class:
```
class CharacterPresentation(private val character: Character) {

    fun formattedDescription(): Int {
        return if (character.description.isEmpty()) R.string.description_not_available else R.string.empty_string
    }
}
```
Then in ViewHolder:
```
holder.tvCharacterDescription.text = characterPresentation.formattedDescription()
```

or

2.- Using a Mapper class to transform the Domain Entity model to a UI Layer model.
```
class CharacterMapper(private val character: Character) {
   fun toCharacterItem() : CharacterItem {
      return CharacterItem(
          name = character.name,
          description = getDescription()
      )
   }
   
   private fun getDescription() : Int {
       if (character.description.isEmpty()) context.getString(R.string.description_not_available) else character.description
   }
}
```
Then in ViewHolder:
```
holder.tvCharacterDescription.text = characterItem.getDescription()
```

Any of these options will also allow you to increase code coverage by adding Unit Tests that verify the appropriate Description is used when it's empty or not.

File: app/src/main/java/com/wizeline/heroes/ui/fragments/CharactersFragment.kt [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+        }
+    }
+
+    private fun refreshCharacters(offset: Int = 0) {
+        if(isInternetAvailable(context!!)){
+            mOffset += offset
+            mMarvelViewModel.getCharacters(mRetrofitRepository, mOffset)
+        } else{
+            changeLoadingState(false)
+            Toast.makeText(context, "No internet connection, please connect to a internet Network", Toast.LENGTH_LONG).show()
```
Comment (pestrada) [inline app/src/main/java/com/wizeline/heroes/ui/fragments/CharactersFragment.kt]: Message should be a string XML resource instead for allowing reuse and language translations (if necessary)

File: app/src/main/java/com/wizeline/heroes/viewmodels/MarvelViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+            .subscribeOn(Schedulers.io())
+            .subscribe({
+                val currentViewState = marvelViewState.value!!
+                val marvelCharacterList = ArrayList<Character>()
+                it.data.results.forEach {
+                    marvelCharacterList.add(it)
+                }
+                currentViewState.characterList = marvelCharacterList
+                currentViewState.isListUpdated = true
+                _marvelViewState.value = currentViewState
```
Comment (pestrada) [inline app/src/main/java/com/wizeline/heroes/viewmodels/MarvelViewModel.kt]: I think most of this boilerplate updates can be replaced by creating an instance directly and remove any `currentViewState` variable. Since `marvelViewState` should be always the current state to render, there's no need for an additional variable. For example we could do it like this:
```
_marvelViewState.value = MarvelViewState(
    characterList = it.data.results,
    isListUpdated = true
)
```
Same with other references to update `marvelViewState.value`. We just pass around the required data through the instance constructor. Also, that way you will avoid the use of `!!` operator.

**Review by pestrada — CHANGES_REQUESTED** [review]
Left a few comments. Code meets requirements for the First Deliverable 🎊 Good job.
Let's just incorporate the suggestions before merging 👍

File: app/build.gradle (hunk 1 above)
Comment (ponchOlvera) [reply app/build.gradle]: Done.

File: app/src/main/java/com/wizeline/heroes/HeroesServices.kt (hunk 2 above)
Comment (ponchOlvera) [reply app/src/main/java/com/wizeline/heroes/HeroesServices.kt]: Done

File: app/src/main/java/com/wizeline/heroes/NetworkClient.kt (hunk 3 above)
Comment (ponchOlvera) [reply app/src/main/java/com/wizeline/heroes/NetworkClient.kt]: Done. created a separate method that creates a new Retrofit instance if the global instance is null and return it. Therefore, it is safe to return the nullable instance as ```retrofitInstance!!```

File: app/src/main/java/com/wizeline/heroes/NetworkClient.kt (hunk 4 above)
Comment (ponchOlvera) [reply app/src/main/java/com/wizeline/heroes/NetworkClient.kt]: Done.

File: app/src/main/java/com/wizeline/heroes/models/Character.kt (hunk 5 above)
Comment (ponchOlvera) [reply app/src/main/java/com/wizeline/heroes/models/Character.kt]: Done.

File: app/src/main/java/com/wizeline/heroes/models/ComicData.kt (hunk 6 above)
Comment (ponchOlvera) [reply app/src/main/java/com/wizeline/heroes/models/ComicData.kt]: Done.

File: app/src/main/java/com/wizeline/heroes/ui/ExtensionFunctions.kt (hunk 7 above)
Comment (ponchOlvera) [reply app/src/main/java/com/wizeline/heroes/ui/ExtensionFunctions.kt]: Done.

File: app/src/main/java/com/wizeline/heroes/ui/ExtensionFunctions.kt (hunk 8 above)
Comment (ponchOlvera) [reply app/src/main/java/com/wizeline/heroes/ui/ExtensionFunctions.kt]: Done.

File: app/src/main/java/com/wizeline/heroes/ui/abstract/PaginationScrollListener.kt (hunk 9 above)
Comment (ponchOlvera) [reply app/src/main/java/com/wizeline/heroes/ui/abstract/PaginationScrollListener.kt]: Done.

File: app/src/main/java/com/wizeline/heroes/ui/adapters/CharacterRecyclerViewAdapter.kt (hunk 10 above)
Comment (ponchOlvera) [reply app/src/main/java/com/wizeline/heroes/ui/adapters/CharacterRecyclerViewAdapter.kt:24]: Change will be implemented on Second Deliverable.

File: app/src/main/java/com/wizeline/heroes/ui/adapters/CharacterRecyclerViewAdapter.kt (hunk 11 above)
Comment (ponchOlvera) [reply app/src/main/java/com/wizeline/heroes/ui/adapters/CharacterRecyclerViewAdapter.kt]: Done.

File: app/src/main/java/com/wizeline/heroes/ui/adapters/CharacterRecyclerViewAdapter.kt (hunk 12 above)
Comment (ponchOlvera) [reply app/src/main/java/com/wizeline/heroes/ui/adapters/CharacterRecyclerViewAdapter.kt]: Done. I sued the Mapper method.

File: app/src/main/java/com/wizeline/heroes/ui/fragments/CharactersFragment.kt (hunk 13 above)
Comment (ponchOlvera) [reply app/src/main/java/com/wizeline/heroes/ui/fragments/CharactersFragment.kt]: Done.

File: app/src/main/java/com/wizeline/heroes/viewmodels/MarvelViewModel.kt (hunk 14 above)
Comment (ponchOlvera) [reply app/src/main/java/com/wizeline/heroes/viewmodels/MarvelViewModel.kt]: Done.

**Review by pestrada — APPROVED** [review]
Good job 👍
